### DISCUSSION TITLE
Help diagnosing the cause of FastAPI application using too much memory

### DISCUSSION THREAD

**Discussion by accountForIssues** [fastapi/fastapi#9857] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# none
```

### Description

Basically, I am migrating my current backend API to Pydantic v2 (with FastAPI v0.100). Currently, the production instance uses about 1 GB of RAM on average with 8 uvicorn workers (started using command line).

After following the migration guide and fixing any errors that came up, I tested the dev instance with 4 workers and the memory use climbed to 2+ GB on start-up with no incoming requests.

I don't have much experience with profiling but I installed memray and ran the application with it and then exported the flamegraph (attached). Looking at that, it appears that either FastAPI (or pydantic) is using a lot of memory to instantiate (response) models.

Note that not very few of the API routes have an explicit response_model and even removing them out didn't reduce the memory use.

Looking at the line number of my code mentioned in the report, nothing stands out so I'm unsure how to proceed forward.

As for the pydantic models, I do have a lot of them with a few being 1-2 level nested ones.

I posted this in pydantic first but after seeing https://github.com/tiangolo/fastapi/discussions/8609, I thought it could be FastAPI related.

Pydantic discussion - https://github.com/pydantic/pydantic/discussions/6550
[memray-flamegraph-uvicorn.1.html.txt](https://github.com/tiangolo/fastapi/files/12021624/memray-flamegraph-uvicorn.1.html.txt)

### Operating System

Linux

### Operating System Details

Debian 12 (Docker)

### FastAPI Version

0.100

### Python Version

3.11.4

**Comment by samuelcolvin** [comment, marked as answer]
Let's discuss in https://github.com/pydantic/pydantic/issues/6620.
